vault/ttl.go

```diff
@@ -0,0 +1,85 @@
+package vault
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+func calculateTTL(sysView logical.SystemView, increment, backendTTL, period, backendMaxTTL, explicitMaxTTL time.Duration, startTime time.Time) (ttl time.Duration, warnings []string, errors error) {
+	// Truncate all times to the second since that is the lowest precision for
+	// TTLs
+	now := time.Now().Truncate(time.Second)
+	if startTime.IsZero() {
+		startTime = now
+	} else {
+		startTime = startTime.Truncate(time.Second)
+	}
+
+	// Use the mount's configured max unless the backend specifies
+	// something more restrictive (perhaps from a role configuration
+	// parameter)
+	maxTTL := sysView.MaxLeaseTTL()
+	if backendMaxTTL > 0 && backendMaxTTL < maxTTL {
+		maxTTL = backendMaxTTL
+	}
+	if explicitMaxTTL > 0 && explicitMaxTTL < maxTTL {
+		maxTTL = explicitMaxTTL
+	}
+
+	// Should never happen, but guard anyways
+	if maxTTL <= 0 {
+		return 0, nil, fmt.Errorf("max TTL must be greater than zero")
+	}
+
+	var maxValidTime time.Time
+	switch {
+	case period > 0:
+		// Cap the period value to the sys max_ttl value
+		if period > maxTTL {
+			warnings = append(warnings,
+				fmt.Sprintf("period of %q exceeded the effective max_ttl of %q; period value is capped accordingly", period, maxTTL))
+			period = maxTTL
+		}
+		ttl = period
+
+		if explicitMaxTTL > 0 {
+			maxValidTime = startTime.Add(explicitMaxTTL)
+		}
+	default:
+		switch {
+		case increment > 0:
+			ttl = increment
+		case backendTTL > 0:
+			ttl = backendTTL
+		default:
+			ttl = sysView.DefaultLeaseTTL()
+		}
+
+		// We cannot go past this time
+		maxValidTime = startTime.Add(maxTTL)
```

Comment:
This switch can be cleaned up a bit if you initialize `maxValidTime := startTime.Add(maxTTL)` and then have the switch like:

```
switch {
	case period > 0:
		...
	case increment > 0:
		ttl = increment
	case backendTTL > 0:
		ttl = backendTTL
	default:
		ttl = sysView.DefaultLeaseTTL()
	}
```

Reply:
The only problem with this is that maxTTL does not limit the period value.  I implemented it this way and it ended up about the same.